File: core/src/bara/game/systems/LightRenderSystem.java
package bara.game.systems;

import com.badlogic.ashley.core.EntitySystem;
import com.badlogic.gdx.graphics.Pixmap;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.glutils.FrameBuffer;
import bara.game.graphics.RenderStage;

public class LightRenderSystem extends EntitySystem implements RenderStage {
    private FrameBuffer _frameBuffer;
    private int _width, _height;

    public LightRenderSystem(int width, int height) {
        super();
        _width = width;
        _height = height;
        _frameBuffer = new FrameBuffer(Pixmap.Format.RGB888, width, height, false);
    }

    @Override
    public void update(float deltaTime) {

    }

    @Override
    public Texture getTexture() {
        return _frameBuffer.getColorBufferTexture();
    }

    @Override
    public void resize(int width, int height) {
        _width = width;
        _height = height;
        _frameBuffer = new FrameBuffer(Pixmap.Format.RGB888, width, height, false);
    }
}

File: core/src/bara/game/systems/LightRenderSystem.kt
package bara.game.systems

import bara.game.ashley.PooledResourceEngine
import com.badlogic.ashley.core.EntitySystem
import bara.game.graphics.RenderStage
import box2dLight.RayHandler
import com.badlogic.ashley.core.Engine
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.graphics.g2d.SpriteBatch

class LightRenderSystem(
    private var width: Int,
    private var height: Int
) : EntitySystem(), RenderStage {

    private lateinit var rayHandler: RayHandler
    private lateinit var gameCamera: OrthographicCamera


    override fun addedToEngine(engine: Engine) {
        if (engine is PooledResourceEngine) {
            rayHandler = engine.getResource<RayHandler>()!!
            rayHandler.setLightMapRendering(false)
            RayHandler.useDiffuseLight(true)       // enable or disable diffused lighting
            rayHandler.setBlur(true)          // enabled or disable blur
            rayHandler.setBlurNum(4)          // set number of gaussian blur passes
            rayHandler.setShadows(true)       // enable or disable shadow
            rayHandler.setCulling(true)       // enable or disable culling
            rayHandler.setAmbientLight(0.9f)  // set default ambient light
            gameCamera = engine.getResource<OrthographicCamera>()!!
        }
    }

    override fun update(deltaTime: Float) {
        rayHandler.setCombinedMatrix(gameCamera)
        Gdx.gl.glClearColor(0f, 0f, 0f, 0f)
        rayHandler.updateAndRender()
        rayHandler.update()
    }

    override fun render(batch: SpriteBatch) {
        batch.draw(
            rayHandler.lightMapTexture,
            0f, 1f,
            1f, -1f
        )
    }

    override fun resize(width: Int, height: Int) {
        this.width = width
        this.height = height
        rayHandler.resizeFBO(width, height)
    }

    override fun dispose() {
    }
}
